# Introduction
Memote reports quite an extensive list of unbalances reactions. Martyn and I have split the list and will look into them by hand to resolve these issues as much as possible.

Here I have 'attached' the file of the total list, and included the split for me and martyn. 

In [1]:
import cameo
import pandas as pd
import cobra.io
import escher

In [2]:
model = cobra.io.read_sbml_model('../model/g-thermo.xml')

In [3]:
unbalanced_rct =["23DK5MPPISO","T_cbl1_e","LARGNHYDRO","ADCL","OCBT","PUNP2","PGK","T_trnaglu_e","PYRACTT","MTRI","R_r0082","PBUTT","T_gdpmann_e","DTPCUNPPT","OMCDC","ABUTD","DHQS","T_udpgalur_e","INO1PPHYDRO","T_adgppuc_e","MDRPD","OXGDC","T_pydx5p_e","KARI_23dhmb","NTD3","T_man_e","T_qh2_3","HBCOAOR","IDPh","ACENLAT","T_trp__L_e","HISOR","DHAD1","T_cyan_e","MHTPPT","T_ile__L_e","T_cmp_e","T_adprib_e","SUCRGT","ACM6PH","OCTACPOR","MOD_3mop","ACGK","T_dhgly_e","ALDD2x","AIZCLY","R_UDPGD","T_pyridoxal_e","T_uri_e","T_phe__L_e","T_ump_e","OCTACPAT","MHOBT","R_SSALx","T_chlac_e","2OH3K5MPPISO","T_abmaagapc_e","EN6LYOR","R_SSALy","ALASm","T_rib__D_e","T_din_e","ENCOAT","NNAM","3OAR100","T_mercplac_e","AMBALY","T_octdecacp_e","T_damp_e","MMM2","ACDO","QUINOR","OPALMACPOR","T_hepdp_e","HDEACPT","DATPT","PUNP1","OAADC","T_gmagg_e","3OAR40_1","ADOCBIP","PTAr","HMBS","XU5PGT","T_but_e","T_gal_e","MALS","ACS","4ABUTD","ATPPT","DHFS","DURIPP","T_co2_e","R_SERD_L","DGTPT","S7PTRANSPP","3HAD40_1","DHORTS","T_betaine_e","URAHL","T_amet_e","T_drib_e","GLNAMHYDRO","T_ps_cho_e","PENTOHYDROARA","PCPPTh","HMSD","R_SELCYSTL","UAAAGGLAAP","T_val__L_e","T_M_10fthf_e","IVCOAR","DHBS","ASAD","T_slnt_e","T_catechol_e","QH2NOR","R_r0068","GCCa","ACMUMPT","PDHam1hi","PHSERT","F1PPT","L23DHDPICOLY","MME","GMPS2","T_accoa_e","ILEOR","PSP","T_phe__D_e","T_his__L_e","ATPNAT","T_etoh_e","GLCDPT","TREPT","ACHMSSELCYSL","NTD5","T_pheme_e","PHEOR","T_gln__D_e","G6PI_1","R_SUCHMSSELCYSL","FTHFLIG","MMSAD1","NTD6","BCFASYN3","IPPS","CHLACOR","AGGPR","T_met__L_e","NTD11","R_UPP1S","PRPAT","MALTPT","CCP","T_a4oxopent_e","T_adn_e","APLh","AMYTRANS","ACBIPGT","GMP5N","G5DHx","STARCHGT","T_fe2_e","T_lac__L_e","T_M_C04051_e","ACAS_2ahbut","FICYTCCNOR","T_prpncoa_e","HC01435R","NO2R","T_amp_e","T_bcfa_e","R_UACMAMO","ADOCBLS","AMETDMMQ","T_oh1_e","ABFPT","COOR","T_csn_e","HMCNSADLY","ARGDC","HGBYR","PUNP3","MANPT","T_uamagdgll_e","MENAOR","G6PI2","PPAPPT","T_M_3hmbcoa_e","T_gmp_e","R_TDP","MAN1PT","QH2OR","T_thmtp_e","NT5C","GAMPT","PRPNTEAT","ACDAH","T_dmbzid_e","T_o2_e","PC20M","R_TALA","UAAAGGLAAT","T_mesel_e","PUNP6","T_fum_e","GLU5K","GDH_nadp","T_chor_e","T_nadph_e","R_TKT2","CELLOHYDRO","T_sprm_e","DECOT","XYLANHYDRO","R_UAGCVT","DECACPOR","T_nadp_e","R_SHSL4h","SELMELY","DECACPAT","T_arg__L_e","GAM","DECDPPP","GDH","T_M_2hymeph_e","ADCPS2","2HBO","T_malcoa_e","T_mmal_e","PUTA3","T_hpglu_e","R_SBP","AUS","DHNAOCTD","ACGS","T_decanoylacp_e","T_fol_e","T_dhptd_e","R_THRPD","SELMELIG","FBAf","UAGUDPGPTRANS","AAGAMPT","R_TMDPPK","DALAOX","T_cpppg1_e","GAPD","R05219","MERCPPYRST","COCHL","T_appl_e","NTD8","GAPDH_nadp_hi","T_thrp_e","ACKr","PUNP4","SCUGLUTRA","XYLHYDRO","PC6YM","HEX7","RAFFINHYDRO","R_VALDHr","R_r0097","GTPCI","LARGNOR","T_cobalt2_e","G5SADrm","R_UPPDC1","NNDPR","GLYALDDr","T_nh4_e","NTD10","MOD","T_lac__D_e","KAS14","FDPPH","AHSL","MMTSAOR","MHOPT","G6PDA","LLEUDr","GDTP","T_for_e","MOD_4mop","T_glc__D_e","T_succ_e","MMALLIG","H202OR","T_acmum_e","ALAALALIG","NP1","R_r0775","PFKh","HACD6","DLDLBT","PYNP1","FPRAMLIG","BFBP","AKP1","CYANST","KARI_23dhmp","SAM","R_r0708","T_ser__L_e","R_SELCYSTGL","FGAMALIG","PUNP7","ICDHhr","PENTOHYDROXYL","KARI","T_selmethtrna_e","PRPUT","ALAALAr","THPPGLIG","IG3PS","T_dgmp_e","PRPGT","CTPS1","AMETTRANS","T_dlglu_c","QULNS","T_mthpglu_e","ACLS","ACCOAC","ASNN","ACSERL","PRPCT","FPGFTh","STACHY","METGL","HOPNTAL","T_M_5fthf_e","T_tcynt_e","CYSAT","ADSL2r","T_dcmp_e","ASP1DC","CTPS2","CHORS","NODOx","AHCYSMETTRANS","T_ftethplg_e","GTPT","CBPS","TRE6PHYDRO","T_cbi_e","T_hqn_e","NODOy","CTPT","FGFTh","T_dtmp_e","MALACPAT","SBZCOAHYDRO","ACYSAT","STACHY2","CYSTGL","T_glyc_e","R_r0163","SALCNPT","MMACPAT","GLYOR","CYSTT","T_tre_e","NAMNPP","FOMETRi","T_ura_e","GLUKA","R_SHSL1","HDH","HCYSMT","T_ala__L_e","STARCHHY","ALKP","PPCDC","FTCL","T_dodecacp_e","CYSDS","PPND","CDPDAGPT","T_dad_2_e","T_M_3omrsACP_e","PYRPT","T_asp__L_e","AHSERL","NMNHYD","T_but2eACP_e","GLYO1","GLYOX","NANEURATRANS","FOCYTCCOR","ADSS","ACPh","DHAD2","ACHBS","T_thf_e","PPCOAOm","HEXACPOR","CYSTHI","KARI_3hmoa","COQOR","HEXACPAT","FDH","R_UGE","BCFASYN","T_dtpcudcpp_e","GND","PPBNGS","ARBTPT","MECDPOR","CBL","PRPNTELIG","PPCOAC","T_glc6P_e","M1PD","2AMACHYD","AHHST","T_gam_e","MNLPT","IPDPCT","R_SHSL2","PAMPT","T_nad_e","PUNP5","PSCIT","DTPCUDCPPR","MAN6PI","T_hxan_e","T_M_5mthf_e","ADPGLCGT","PDH","NNDMBRT","GTPCII2","ASAPT","AMYLOSEGT","FBA3","MGSA","ANS","T_nadh_e","T_dcyt_e","DDPA","PYDXS","R_r0120","ME1","T_gthrd_e","AMETLY","R_TAL","ECOAH4","ASPO1","T_thymd_e","BGLUCH","T_aacoa_e","T_coa_e","PPCK","DCTPT","MSAOR","R_TMPPP","NTD7","APAPTm","T_inost_e","CYSS","UTPT","MALTHYDRO","T_capra_e","ATPM","DXPS","T_mlthf_e","SUCHY","DAPDC","PC","T_RNA_e","T_ac_e","T_gthox_e","SUCHY2","DTTPT","MMSAD3","NTD1","T_succcoa_c","T_hmgcoa_e","FRU1PT","T_hemeO_e","STARCHHY2","T_g3pc_e","CS","NADS1","CELLBHY","T_acetol_e","ALAD_L","SMBCOAOR","PC17M","T_dna_e","MYRACPOR","T_hdeacp_e","AACOAT","GLUTRR","R_SEPHCHCS","PGCM","DTPCUNPTRANS","T_acon_e","T_uacmamu_e","GLUTRS","OMPDC","MBTCOAT","T_thm_e","T_asn__L_e","T_pmtcoa_e","DTPCUNPT","HBCOAH","T_14dhncoa_e","GCCb","FOCYTCNOR","NTP3","R_UAMAS","HEPDPP","T_ribflv_e","T_arbt_e","NTD2","ASPCT","T_mmcoa__S_e","T_gly_e","BCFASYN2","R_UPP3MT","PPGPPDP","SELCYSLY","AKGDH","GALISO","T_menqui_e","T_salcn_e","T_ins_e","PPNDH","PPRGL","ACSERT","NTD4","R_UAMAGS","DB4PS","R_THRS","ACYP_2","HISTP","IGPS","T_h_e","T_glcn__D_e","ASPK","ASPSALY","ADCPS1","ARABINANHY","UAMAGDGLLLIG","R_SUCBZL","BUTACPAT","AH6PISO","SCENLAT","T_so3_e","ASPO6","PC11M","ADOCBLTRANS","PRAISh","ADOCBIK","MYOIPPH","UAMAGLIG","RZ5PP","FTHFD","FTETHLGPT","T_cl_e","ECOAH9ir","AIRC3","ASNS2","PRE3AOR","GHMT","PANOPTRANS","T_nac_e"]
len(unbalanced_rct)

570

In [4]:
martyn_reactions = unbalanced_rct[0:285]
len(martyn_reactions)

285

In [5]:
vivienne_reactions = unbalanced_rct[286:569]

In [6]:
#make a list of all of Viviennes reactions that exlclude transport
metabolic_rct_viv = []
for rct in model.reactions:
    if rct.id in vivienne_reactions:
        if rct.id[-1:] in 't':
            continue
        else: 
            metabolic_rct_viv.append(rct)
    else: continue

In [7]:
len(metabolic_rct_viv)

190

So there are 190 metabolic reactions that I need to check.

I can see from inspecting some of the below reactions that the majority of the imbalance comes from issues with protons. After talking with Ben and Nikolaus Sonnenschein, he mentioned that the best option would be to fix the mass imbalance by leaving the protonation state of each metabolite as it is (at physiological pH) and instead balancing out the reactions by adding protons to the reaction. 

So to tackle this, I will try to write a script, where for each unbalanced reaction, it will check the balance of hydrogens and correspondingly add H+ to the reaction where needed. If the imbalance comes from something other than the protons, i will store it into a nother list. Later I can run memote and check the two lists again one another to make sure that the reactions are now balanced with hydrogen, and then go further to see what to do with the other reactions.


In [8]:
#make a list of all unbalanced metabolic reactions
metabolic_rct = []
for rct in model.reactions:
    if rct.id in unbalanced_rct:
        if rct.id[-1:] in 't':
            continue
        else: 
            metabolic_rct.append(rct)
    else: continue

In [9]:
len(metabolic_rct)

364

In [11]:
#calc Number of each element in the metabolite in this reaction
C_unbalanced =[]
O_unbalanced =[]
N_unbalanced =[]
P_unbalanced =[]
error_formula = []

for rct in metabolic_rct:
#make one dictionary that contains the information about the total elements involved in each metabolite of the whole reaction
    rct_dict = {}
    all_H = []
    all_C = []
    all_O = []
    all_N = []
    all_P = []
    stoich = {}
    for met in rct.metabolites.keys():
        met_formula = met.elements
        if not met.elements:
            error_formula.append(met)
        else:
            try:
                H = met_formula['H']
            except KeyError:
                H = 0
            try:
                C = met_formula['C']
            except KeyError:
                C = 0
            try: 
                O = met_formula['O']
            except KeyError:
                O = 0
            try: 
                N = met_formula['N']
            except KeyError:
                N = 0
            try: 
                P = met_formula['P']
            except KeyError:
                P = 0
            elements = {f"H_{met.id}":H,f"C_{met.id}":C, f"O_{met.id}":O,f"N_{met.id}":N,f"P_{met.id}":P}
            rct_dict.update(elements)
#make a dictionary that has the information about the met.id and stoichiometry
        for met in rct.metabolites.keys():
            stoichiometry = rct.metabolites.get(met)
            stoich[met.id]=stoichiometry
        #make a calculation for the final balance of hydrogen for example:
    for key in rct_dict:
        id_key = stoich[key[2:]]
        el = key[:2]
        if el in 'H_':
            num_H = rct_dict.get(key)
            totH_key = num_H * id_key
            all_H.append(totH_key)
        elif el in 'C_':
            num_C = rct_dict.get(key)
            totC_key = num_C * id_key
            all_C.append(totC_key)
        elif el in 'O_':
            num_O = rct_dict.get(key)
            totO_key = num_O * id_key
            all_O.append(totO_key)
        elif el in 'N_':
            num_N = rct_dict.get(key) 
            totN_key = num_N * id_key
            all_N.append(totN_key)
        elif el in 'P_':
            num_P = rct_dict.get(key)
            totP_key = num_P * id_key
            all_P.append(totP_key)
        else: 
            continue
#add the number of that element to the list, and sum all parts of the list. this sum should equal 0 or it is unbalanced
    balance_H = sum(all_H) 
    balance_C = sum(all_C)
    balance_O = sum(all_O)
    balance_N = sum(all_N)
    balance_P = sum(all_P)
    if balance_H != 0 and balance_C == 0 and balance_O == 0:
    #Make a script to balance the reaction with a proton)
        add_H = -sum(all_H) #i.e. the stoichiometry of the proton needed to get mass balance
        proton = model.metabolites.h_c
        rct.add_metabolites({proton:add_H})
    if balance_C!= 0:
        C_unbalanced.append(rct)
    if balance_O != 0:
        O_unbalanced.append(rct)
    if balance_N != 0:
        N_unbalanced.append(rct)
    if balance_P != 0:
        P_unbalanced.append(rct)
   

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in 'C10H18O13P2R2(C5H8O6PR)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in '(C12H20O10)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in 'C14H18O4(C5H8)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in 'C20H30N6O11PSeR(C5H8O6PR)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in '(C6H10O5)n'



In [24]:
print('Number of carbon unbalanced rcts:', len(C_unbalanced))
print('Number of oxygen unbalanced rcts:', len(O_unbalanced))
print('Number of nitrogen unbalanced rcts:', len(N_unbalanced))
print('Number of phosphorous unbalanced rcts:', len(P_unbalanced))

Number of carbon unbalanced rcts: 71
Number of oxygen unbalanced rcts: 89
Number of nitrogen unbalanced rcts: 44
Number of phosphorous unbalanced rcts: 50


In [30]:
#Find intersection between all the lists
overlay = list(set(C_unbalanced) & set(O_unbalanced) & set(N_unbalanced) & set(P_unbalanced))
len(overlay)

26

In [53]:
#above coding made the error_formula list contain duplicate (or more) of reactions, so we need to get rid of these.
error_formula_dict = list(dict.fromkeys(error_formula))

In [54]:
len(error_formula_dict)

9

In [46]:
#make list of all reactions imbalanced by C, N, O, P or a combination of any
total_unbalanced = []
for rct in model.reactions:
    if rct in C_unbalanced:
        total_unbalanced.append(rct)
    if rct in O_unbalanced:
        if rct in total_unbalanced:
            continue
        else:
            total_unbalanced.append(rct)
    if rct in N_unbalanced and not total_unbalanced:
        if rct in total_unbalanced:
            continue
        else:
            total_unbalanced.append(rct)
    if rct in P_unbalanced and not total_unbalanced:
        if rct in total_unbalanced:
            continue
        else:
            total_unbalanced.append(rct)

In [48]:
len(total_unbalanced)

93

So there are 26 reactions with all the elements unbalanced (overlay), and 9 metabolites which have an error in the formula (error_formula). These 27 may be the cause for the unbalanced reactions. 
You could fix these 27 formulas and re-run the code with the left over unbalanced reactions to see if this solves a majority of the problems.

Overall, there are 93 reactions left with an inbalance from either one or more of the C, N, O and P elements. These will need to be investigated on a case by case basis.



In [49]:
#save&commit and check in memote if mass balance really improved
cobra.io.write_sbml_model(model, '../model/g-thermo.xml')

## Fixing metabolites with erroring formula
In the dictionary [error_formula_dict] we've stored metabolites that give an error because they contain some parenthesis. Here I will fix those by hand and then re-run the above code to see if this gets rid of some of the imbalanced reactions we identified above.

To tackle the errored formulas, I will just go through them one by one here by hand to modify them, as there are only 9.



Once that is done, I can send the list of all imbalanced reactions still left to Martyn so that we can divide the tasks of which need to be done by hand. 

In [102]:
model = cobra.io.read_sbml_model('../model/g-thermo.xml')

In [111]:
model.metabolites.gmagg_c       

<Metabolite gmagg_c at 0x1dca85ffdc8>

In [112]:
model.metabolites.gmagg_c.notes

{'CHARGE': '0',
 'KEGG': 'C11826',
 'NAME': '[GlcNAc-(1⇢4)-Mur2Ac(oyl-L-Ala-g-D-Glu-L-Lys-D-Ala-D-Ala)]n+1-diphosphoundecaprenol; [GlcNAc-(1⇢4)-Mur2Ac(oyl-L-Ala-gamma-D-Glu-L-Lys-D-Ala-D-Ala)]n-diphosphoundecaprenol; [GlcNAc-(1⇢4)-Mur2Ac(oyl-L-Ala-g-D-Glu-L-Lys-D-Ala-D-Ala)]n-diphosphoundecaprenol; [GlcNAc-(1⇢4)-Mur2Ac(oyl-L-Ala-gamma-D-Glu-L-Lys-D-Ala-D-Ala)]n+1-diphosphoundecaprenol',
 'ChEBI': '37790'}

In [105]:
model.metabolites.RNA_c.formula = 'C15H24O18P3R3'

In [106]:
model.metabolites.ubiquin_c.formula = 'C19H26O4R'

In [107]:
model.metabolites.edhyl_c.formula = 'C8H16NOS2R'

In [108]:
model.metabolites.smbdhlp_c.formula = 'C13H24NO2S2R'

In [109]:
model.metabolites.focytB561_c.formula = 'R' 
#not really ideal, this reaction needs to be checked anyway. will get back to later.

In [110]:
model.metabolites.selmethtrna_c.formula = 'C25H38N6O17P2SeR2'

Note for starch and cellulose it is difficult to properly assign a formula. We will discuss this and come back to fixing them another time. (this is also the case for gmagg_c as it is also a glycan)

In [100]:
model.reactions.BGLUCH

<Reaction BGLUCH at 0x1dca6441d48>

In [57]:
error_formula_dict

[<Metabolite RNA_c at 0x1dca0a1c488>,
 <Metabolite starch_c at 0x1dca39d6088>,
 <Metabolite ubiquin_c at 0x1dca39c0288>,
 <Metabolite edhyl_c at 0x1dca0a1c108>,
 <Metabolite smbdhlp_c at 0x1dca0a16748>,
 <Metabolite focytB561_c at 0x1dca39dc1c8>,
 <Metabolite selmethtrna_c at 0x1dca0a16588>,
 <Metabolite cellulose_c at 0x1dca39bd388>,
 <Metabolite gmagg_c at 0x1dca39ec088>]

In [119]:
#calc Number of each element in the metabolite in this reaction
C_unbalanced =[]
O_unbalanced =[]
N_unbalanced =[]
P_unbalanced =[]
error_formula = []

for rct in metabolic_rct:
#make one dictionary that contains the information about the total elements involved in each metabolite of the whole reaction
    rct_dict = {}
    all_H = []
    all_C = []
    all_O = []
    all_N = []
    all_P = []
    stoich = {}
    for met in rct.metabolites.keys():
        met_formula = met.elements
        if not met.elements:
            error_formula.append(met)
        else:
            try:
                H = met_formula['H']
            except KeyError:
                H = 0
            try:
                C = met_formula['C']
            except KeyError:
                C = 0
            try: 
                O = met_formula['O']
            except KeyError:
                O = 0
            try: 
                N = met_formula['N']
            except KeyError:
                N = 0
            try: 
                P = met_formula['P']
            except KeyError:
                P = 0
            elements = {f"H_{met.id}":H,f"C_{met.id}":C, f"O_{met.id}":O,f"N_{met.id}":N,f"P_{met.id}":P}
            rct_dict.update(elements)
#make a dictionary that has the information about the met.id and stoichiometry
        for met in rct.metabolites.keys():
            stoichiometry = rct.metabolites.get(met)
            stoich[met.id]=stoichiometry

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in 'C10H18O13P2R2(C5H8O6PR)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in '(C12H20O10)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in 'C14H18O4(C5H8)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in 'C20H30N6O11PSeR(C5H8O6PR)n'

C:\Users\vivmol\AppData\Local\Continuum\anaconda3\envs\g-thermo\lib\site-packages\cobra\core\metabolite.py:89: UserWarning:

invalid formula (has parenthesis) in '(C6H10O5)n'



In [121]:
#make a calculation for the final balance of hydrogen for example:
all_H = []
all_C = []
all_O = []
all_N = []
all_P = []
for key in rct_dict:
    id_key = stoich[key[2:]]
    el = key[:2]
    if el in 'H_':
        num_H = rct_dict.get(key)
        totH_key = num_H * id_key
        all_H.append(totH_key)
    elif el in 'C_':
        num_C = rct_dict.get(key)
        totC_key = num_C * id_key
        all_C.append(totC_key)
    elif el in 'O_':
        num_O = rct_dict.get(key)
        totO_key = num_O * id_key
        all_O.append(totO_key)
    elif el in 'N_':
        num_N = rct_dict.get(key) 
        totN_key = num_N * id_key
        all_N.append(totN_key)
    elif el in 'P_':
        num_P = rct_dict.get(key)
        totP_key = num_P * id_key
        all_P.append(totP_key)
    else: 
        continue
#add the number of that element to the list, and sum all parts of the list. this sum should equal 0 or it is unbalanced
balance_H = sum(all_H) 
balance_C = sum(all_C)
balance_O = sum(all_O)
balance_N = sum(all_N)
balance_P = sum(all_P)
if balance_H != 0 and balance_C == 0 and balance_O == 0:
    #Make a script to balance the reaction with a proton)
    add_H = -sum(all_H) #i.e. the stoichiometry of the proton needed to get mass balance
    proton = model.metabolites.h_c
    rct.add_metabolites({proton:add_H})
if balance_C!= 0:
    C_unbalanced.append(rct)
if balance_O != 0:
    O_unbalanced.append(rct)
if balance_N != 0:
    N_unbalanced.append(rct)
if balance_P != 0:
    P_unbalanced.append(rct)
   

In [77]:
rct

<Reaction PPBNGS at 0x2252e55ad08>

After automatic fix: save and check in memote that mass balance significantly improves.

Later: fix and inspect the error_formula metabolites
and check all the C, O, N and P unbalanced reactions